src/common.c

```diff
@@ -733,6 +745,168 @@ retrace_print_parameter(struct rtr_print_buf *print_buffer, unsigned int event_t
+			break;
+
+		default:
+			break;
+		}
+
+		trace_printf(print_buffer, 0, "%s(%d)", domain_str, domain);
+		break;
+	}
+
+	case PARAMETER_TYPE_SOCKET_TYPE:
+	{
+		int type = (*(int *) *value);
+		const char *type_str = "SOCKET_TYPE";
+
+		switch (type) {
+		case SOCK_STREAM:
+			type_str = "SOCK_STREAM";
+			break;
+
+		case SOCK_DGRAM:
+			type_str = "SOCK_DGRAM";
+			break;
+
+		case SOCK_SEQPACKET:
+			type_str = "SOCK_SEQPACKET";
+			break;
+
+		case SOCK_RAW:
+			type_str = "SOCK_RAW";
+			break;
+
+#if HAVE_DECL_SOCK_PACKET
+		case SOCK_PACKET:
+			type_str = "SOCK_PACKET";
+			break;
+#endif
+
+		default:
+			break;
+		}
+
+		trace_printf(print_buffer, 0, "%s(%d)", type_str, type);
+		break;
+	}
+
+	case PARAMETER_TYPE_SOCKET_OPTION:
+	{
+		int i, opt = (*(int *) *value);
+		const char *opt_str = "SOCKET_OPTION";
+
+		struct socket_opt_data {
+			int opt;
+			const char *opt_str;
+		} socket_opts[] = {
+			/* socket level options */
+			{SO_ACCEPTCONN, "SO_ACCEPTCONN"}, {SO_BROADCAST, "SO_BROADCAST"},
+			{SO_DEBUG, "SO_DEBUG"},	{SO_ERROR, "SO_ERROR"},
+#if HAVE_DECL_SO_BINDTODEVICE
+			{SO_BINDTODEVICE, "SO_BINDTODEVICE"},
+#endif
+#if HAVE_DECL_SO_BSDCOMPAT
+			{SO_BSDCOMPAT, "SO_BSDCOMPAT"},
+#endif
+#if HAVE_DECL_SO_DOMAIN
+			{SO_DOMAIN, "SO_DOMAIN"},
+#endif
+#if HAVE_DECL_SO_MARK
+			{SO_MARK, "SO_MARK"},
+#endif
+#if HAVE_DECL_SO_PASSCRED
+			{SO_PASSCRED, "SO_PASSCRED"},
+#endif
+#if HAVE_DECL_SO_PEERCRED
+			{SO_PEERCRED, "SO_PEERCRED"},
+#endif
+#if HAVE_DECL_SO_PRIORITY
+			{SO_PRIORITY, "SO_PRIORITY"},
+#endif
+#if HAVE_DECL_SO_PROTOCOL
+			{SO_PROTOCOL, "SO_PROTOCOL"},
+#endif
+#if HAVE_DECL_SO_RXQ_OVFL
+			{SO_RXQ_OVFL, "SO_RXQ_OVFL"},
+#endif
+#if HAVE_DECL_SO_SNDBUFFORCE
+			{SO_SNDBUFFORCE, "SO_SNDBUFFORCE"},
+#endif
+#if HAVE_DECL_SO_RCVBUFFORCE
+			{SO_RCVBUFFORCE, "SO_RCVBUFFORCE"},
+#endif
+#if HAVE_DECL_SO_BUSY_CALL
+			{SO_BUSY_POLL, "SO_BUSY_POLL"},
+#endif
+			{SO_DONTROUTE, "SO_DONTROUTE"}, {SO_KEEPALIVE, "SO_KEEPALIVE"},
+			{SO_LINGER, "SO_LINGER"}, {SO_OOBINLINE, "SO_OOBINLINE"},
+			{SO_RCVBUF, "SO_RCVBUF"}, {SO_RCVLOWAT, "SO_RCVLOWAT"}, {SO_SNDLOWAT, "SO_SNDLOWAT"},
+			{SO_RCVTIMEO, "SO_RCVTIMEO"}, {SO_SNDTIMEO, "SO_SNDTIMEO"},
+			{SO_REUSEADDR, "SO_REUSEADDR"}, {SO_REUSEPORT, "SO_REUSEPORT"},
+			{SO_SNDBUF, "SO_SNDBUF"}, {SO_TIMESTAMP, "SO_TIMESTAMP"}, {SO_TYPE, "SO_TYPE"},
+
+			/* IPv6 level options */
+			{IPV6_JOIN_GROUP, "IPV6_JOIN_GROUP"}, {IPV6_LEAVE_GROUP, "IPV6_LEAVE_GROUP"},
+			{IPV6_MULTICAST_HOPS, "IPV6_MULTICAST_HOPS"}, {IPV6_MULTICAST_IF, "IPV6_MULTICAST_IF"},
+			{IPV6_MULTICAST_LOOP, "IPV6_MULTICAST_LOOP"}, {IPV6_UNICAST_HOPS, "IPV6_UNICAST_HOPS"},
+			{IPV6_V6ONLY, "IPV6_V6ONLY"}, {IPV6_RECVTCLASS, "IPV6_RECVTCLASS"},
+			{IPV6_TCLASS, "IPV6_TCLASS"},
+
+			{-1, NULL}
+		};
+
+		for (i = 0; socket_opts[i].opt_str != NULL; i++) {
+			if (socket_opts[i].opt == opt)
+				opt_str = socket_opts[i].opt_str;
+		}
+
+		trace_printf(print_buffer, 0, "%s(%d)", opt_str, opt);
```

Comment:
I think a test for non-null `opt_str` would be sensible otherwise this code relies on undefined `printf` behaviour for NULL pointer

Reply:
@drystone `opt_str` was initialized at declaration as `SOCKET_OPTION`, but I will try to add the checking statement whether `opt` has `-1`